addons/mail/controllers/mail.py

```diff
@@ -165,3 +166,31 @@ def mail_action_view(self, model=None, res_id=None, access_token=None, **kwargs)
             except ValueError:
                 res_id = False
         return self._redirect_to_record(model, res_id, access_token, **kwargs)
+
+    # csrf is disabled here because it will be called by the MUA with unpredictable session at that time
+    @http.route('/mail/unfollow', type='http', auth='public', csrf=False)
+    def mail_action_unfollow(self, model, res_id, pid, token, **kwargs):
+        comparison, record, __ = MailController._check_token_and_record_or_redirect(model, int(res_id), token)
+        if not comparison or not record:
+            raise AccessError(_('Non existing record or wrong token.'))
+
+        pid = int(pid)
+        record_sudo = record.sudo()
+        record_sudo.message_unsubscribe([pid])
+
+        # Check rights as internal user can unfollow without read access and rights could have been revoked
+        has_access = False
+        records_with_user = [record] if request.session.uid \
```

Comment:
Actually, a partner can be linked to multiple user. Here we check that the partner has at least one user that has the read right on the record if the user is not connected. The idea is:
- if the user is connected, we display the link only if he has access to the record with that user. So that if it click on the link there will be no error.
- if the user is not connected, we check that at least one user of the partner has access to it. So when he clicks on the link, he can log in with that user and display the record.

So I think we cannot simplify it. What do you think ?

Reply:
Ah ok ... what about a flow a bit different, like

  * you are logged -> check access, if yes, redirect to record directly (no portal page, actually not necessary), if no, display portal page
  * you are not logged -> display portal page

And we remove the record link in portal page. Or we always display the link, and let the user login if necessary (whatever its rights, those will be managed by the /view controller).

Reply:
Ah ok ... what about a flow a bit different, like

  * you are logged -> check access, if yes, redirect to record directly (no portal page, actually not necessary), if no, display portal page
  * you are not logged -> display portal page

And we remove the record link in portal page. Or we always display the link, and let the user login if necessary (whatever its rights, those will be managed by the /view controller).

Reply:
Actually, we have chosen to display that page (You no longer follow the record...) because displaying a notification after redirecting to the record page what not easy and we wanted to notify the user that the unfollow action has worked. So the link is required.

To simplify the code, we could always display the link as you suggest. At least with that the user will know that he has not access to it (after clicking on it). I will do that and check with Luc (with another think I have noticed).

(the most user friendly version would be to directly display that the user has not access to the record)

Reply:
Oh yes, I remember. Then always display the link for non logged users, and for logged users, depend on their actual ACLs. That way we don't have to try all users of a partner_id, I prefer to avoid doing some guess game with unknown people :)